source/extensions/common/wasm/wavm/wavm.cc

```diff
@@ -585,6 +620,36 @@ template void getFunctionWavm<uint32_t, uint32_t, uint32_t, uint32_t, uint32_t,
     std::function<uint32_t(Context*, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t,
                            uint32_t, uint32_t, uint32_t, uint32_t)>*);
 
+template <typename T> T getValue(IR::Value) {}
+template <> int32_t getValue(IR::Value v) { return v.i32; }
+template <> uint32_t getValue(IR::Value v) { return v.u32; }
+template <> int64_t getValue(IR::Value v) { return v.i64; }
+template <> uint64_t getValue(IR::Value v) { return v.u64; }
+template <> float getValue(IR::Value v) { return v.f32; }
+template <> double getValue(IR::Value v) { return v.f64; }
+
+template <typename T> T WavmGlobal<T>::get() {
+  return getValue<T>(getGlobalValue(wavm_->context_, global_));
+}
+
+template <typename T> void WavmGlobal<T>::set(const T& t) {
+  setGlobalValue(wavm_->context_, global_, IR::Value(t));
+}
+
+template <typename T>
+std::unique_ptr<Global<T>> makeGlobalWavm(WasmVm* vm, absl::string_view moduleName,
+                                          absl::string_view name, T initialValue) {
+  auto wavm = static_cast<Common::Wasm::Wavm::Wavm*>(vm);
+  auto g =
```

Comment:
Why are you doing this? You may always write as below. There is no penalty and it is exception safe.

```
auto g_ptr = std::make_unique<WavmGlobal<T>>(...);
wavm->intrinsicGlobals_[...] = g_ptr.get();
return g_ptr;
```

Reply:
It is a bit more complicated because the return type is std::unique_ptr<Global<T>>> which the compiler will not auto-coerce from 'g'.  Instead I have to do a std::static_pointer_cast<> and this works around that by using the automatic std::unique_ptr constructor coercion which does upcast "correctly" (IMO std::unique_ptr should do that was well, maybe in a later version of C++).

I'll change it.

Reply:
My bad.  You are right.  The problem only happens with shared_ptr.  Done.